syncstorage/storage/spanner.py

```diff
@@ -0,0 +1,988 @@
+                "SELECT CURRENT_TIMESTAMP()"
+            ).one()[0]
+            session.timestamp = dt2ts(result)
+            session.collection_ts = 0
+
+    @contextlib.contextmanager
+    def lock_for_read(self, ruser, collection):
+        with SpannerStorageSession(self) as spanner_session:
+            with self._database.snapshot(multi_use=True) as txn:
+                spanner_session.set_transaction(txn)
+                yield None
+
+    @contextlib.contextmanager
+    def lock_for_write(self, user, collection):
+        with SpannerStorageSession(self) as spanner_session:
+            collectionid = self._get_collection_id(collection, True)
+            with SessionCheckout(self._pool) as session:
+                with session.transaction() as transaction:
+                    spanner_session.set_transaction(transaction)
+                    self._lock_write_precondition(
+                        spanner_session,
+                        user,
+                        collectionid
+                    )
+                    try:
+                        yield None
+                    except Aborted:
+                        raise ConflictError
+
+    #
+    # APIs to operate on the entire storage.
+    #
+
+    @with_session
+    def get_storage_timestamp(self, session, user):
+        """Returns the storage timestamp for the user"""
+        result = session.transaction.execute_sql(
+            getq(queries.STORAGE_TIMESTAMP),
+            params={"userid": user_key(user)},
+            param_types={"userid": param_types.STRING}
+        ).one()[0]
+        return dt2ts(result) if result else 0
+
+    @with_session
+    def get_collection_timestamps(self, session, user):
+        """Returns the collection timestamps for a user."""
+        userid = user_key(user)
+        res = session.transaction.execute_sql(
+            getq(queries.COLLECTIONS_TIMESTAMPS),
+            params={"userid": userid},
+            param_types={"userid": param_types.STRING}
+        )
+        res = self._map_collection_names(res)
+        for collection in res:
+            res[collection] = dt2ts(res[collection])
+        return res
+
+    @with_session
+    def get_collection_counts(self, session, user):
+        """Returns the collection counts."""
+        userid = user_key(user)
+        res = session.transaction.execute_sql(
+            COLLECTIONS_COUNTS,
+            params={"userid": userid},
+            param_types={"userid": param_types.STRING}
+        )
+        return self._map_collection_names(res)
+
+    @with_session
+    def get_collection_sizes(self, session, user):
+        """Returns the total size for each collection."""
+        userid = user_key(user)
+        res = session.transaction.execute_sql(
+            COLLECTIONS_SIZES,
+            params={"userid": userid},
+            param_types={"userid": param_types.STRING}
+        )
+        # Some db backends return a Decimal() instance for this aggregate.
```

Comment:
Does "some db backends" still apply in this case where you have a single known db backend?

Reply:
Heh, copypasta gone awry.